node/src/accountant/big_int_db_processor.rs

```diff
@@ -0,0 +1,1613 @@
+                            0
+                        } + wei_change_high_b_component
+                            + 1
+                    )
+                }
+                let single_numbered_balance = BigIntDivider::reconstitute(high_bytes, low_bytes);
+                assert_eq!(single_numbered_balance, expected_sum);
+                Ok(())
+            })
+            .unwrap();
+    }
+
+    fn precise_upsert_or_update_assertion(
+        test_name: &str,
+        main_sql: &str,
+        update_clause_opt: Option<for<'a> fn(&'a str) -> String>,
+        init_record_opt: Option<(&str, i64, i64)>,
+        requested_wei_change: WeiChange,
+        do_we_expect_overflow: bool,
+    ) {
+        let act = |conn: &mut dyn ConnectionWrapper, subject: &BigIntDbProcessor<DummyDao>| {
+            subject.execute(
+                Left(conn),
+                BigIntSqlConfig::new(
+                    main_sql,
+                    update_clause_opt,
+                    SQLParamsBuilder::default()
+                        .key("name", ":name", &"Joe")
+                        .wei_change(requested_wei_change.clone())
+                        .build(),
+                ),
+            )
+        };
+
+        precise_upsert_or_update_assertion_test_environment(
+            test_name,
+            init_record_opt,
+            &requested_wei_change,
+            do_we_expect_overflow,
+            act,
+        )
+    }
+
+    fn precise_upsert_or_update_assertion_test_environment<F>(
+        test_name: &str,
+        init_record_opt: Option<(&str, i64, i64)>,
+        requested_wei_change: &WeiChange,
+        do_we_expect_overflow: bool,
+        act: F,
+    ) where
+        F: Fn(
+            &mut dyn ConnectionWrapper,
+            &BigIntDbProcessor<DummyDao>,
+        ) -> Result<(), BigIntDbError>,
+    {
+        let mut conn = initiate_simple_connection_and_test_table("big_int_db_processor", test_name);
+        if let Some(values) = init_record_opt {
+            insert_single_record(conn.as_ref(), [&values.0, &values.1, &values.2])
+        };
+        let subject = BigIntDbProcessor::<DummyDao>::default();
+
+        let result = act(conn.as_mut(), &subject);
+
+        assert_eq!(result, Ok(()));
+        let wei_change = match requested_wei_change {
+            Addition(_, num) => checked_conversion::<u128, i128>(*num),
+            Subtraction(_, num) => checked_conversion::<u128, i128>(*num).neg(),
+        };
+        let wei_change_deconstructed = BigIntDivider::deconstruct(wei_change);
+        let (_, low_bytes) = wei_change_deconstructed;
+        let did_overflow_occurred = if let Some(values) = init_record_opt {
+            values.2
+        } else {
+            0
+        }
+        .checked_add(low_bytes)
+        .is_none();
+        assert_eq!(
+            did_overflow_occurred, do_we_expect_overflow,
+            "we encountered overflow on low bytes thought hadn't been expected"
```

Comment:
Message could be improved...maybe "Expected overflow: true; encountered overflow: false"